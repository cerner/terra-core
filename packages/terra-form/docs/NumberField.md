# Number Field

Number field component for rendering text inputs with standard structured style defaulted to type=number. Also exposes attributes specific to that input type. Uses the input component.

## Usage

```jsx
import React from 'react';
import { NumberField } from 'terra-form';

<NumberField
  label="Sales Tax Rate"
  name="sales_tax_rate"
  value={0.075}
  error="This field is required"
  help="Your county's office may have this information"
  required={true}
  max={1}
  min={0}
  step={0.1}
  attrs={{ className: 'healtheintent-application' }}
  isInline={false}
/>
```
