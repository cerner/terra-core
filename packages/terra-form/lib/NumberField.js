'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Field = require('./Field');

var _Field2 = _interopRequireDefault(_Field);

var _Input = require('./Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  /**
   * The defaultValue of the input element
   */
  defaultValue: _propTypes2.default.number,
  /**
   * Error message for when the input is invalid
   */
  error: _propTypes2.default.node,
  /**
   *  Help element to display with the input
   */
  help: _propTypes2.default.node,
  /**
   * Additional attributes to attach to the number input
   */
  inputAttrs: _propTypes2.default.object,
  /**
   * Whether the field is inline
   */
  isInline: _propTypes2.default.bool,
  /**
   * Label of the input
   */
  label: _propTypes2.default.node,
  /**
   * Maximum value allowed for the input
   */
  max: _propTypes2.default.number,
  /**
   * Minimum value allowed for the input
   */
  min: _propTypes2.default.number,
  /**
   * Name of the input attribute
   */
  name: _propTypes2.default.string,
  /**
   * Function to trigger when user changes the input value
   */
  onChange: _propTypes2.default.func,
  /**
   * Whether the input is required
   */
  required: _propTypes2.default.bool,
  /**
   * Incremental steps for increasing and descreasing the value
   */
  step: _propTypes2.default.number,
  /**
   * The value of the input element
   */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

var defaultProps = {
  defaultValue: undefined,
  error: null,
  help: null,
  inputAttrs: {},
  isInline: false,
  label: null,
  max: null,
  min: null,
  name: null,
  onChange: undefined,
  required: true,
  step: null,
  value: undefined
};

var NumberField = function NumberField(_ref) {
  var error = _ref.error,
      help = _ref.help,
      inputAttrs = _ref.inputAttrs,
      isInline = _ref.isInline,
      label = _ref.label,
      max = _ref.max,
      min = _ref.min,
      name = _ref.name,
      onChange = _ref.onChange,
      required = _ref.required,
      step = _ref.step,
      value = _ref.value,
      defaultValue = _ref.defaultValue,
      customProps = _objectWithoutProperties(_ref, ['error', 'help', 'inputAttrs', 'isInline', 'label', 'max', 'min', 'name', 'onChange', 'required', 'step', 'value', 'defaultValue']);

  return _react2.default.createElement(
    _Field2.default,
    _extends({
      error: error,
      label: label,
      help: help,
      isInline: isInline,
      required: required
    }, customProps),
    _react2.default.createElement(_Input2.default, _extends({
      type: 'number',
      step: step,
      max: max,
      min: min,
      required: required,
      name: name,
      value: value,
      defaultValue: defaultValue,
      onChange: onChange
    }, inputAttrs))
  );
};

NumberField.propTypes = propTypes;
NumberField.defaultProps = defaultProps;

exports.default = NumberField;