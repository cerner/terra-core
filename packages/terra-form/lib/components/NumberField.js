'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Field = require('./Field');

var _Field2 = _interopRequireDefault(_Field);

var _Input = require('./Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } // eslint-disable react/jsx-boolean-value

var propTypes = {
  /**
   * Additional Attributes to attach to the number input
   */
  attrs: _react.PropTypes.object,
  /**
   * The defaultValue of the input element
   */
  defaultValue: _react.PropTypes.number,
  /**
   * Error message for when the input is invalid
   */
  error: _react.PropTypes.node,
  /**
   *  Help element to display with the input
   */
  help: _react.PropTypes.node,
  /**
   * Whether the field is inline
   */
  isInline: _react.PropTypes.bool,
  /**
   * Label of the input
   */
  label: _react.PropTypes.node,
  /**
   * Maximum value allowed for the input
   */
  max: _react.PropTypes.number,
  /**
   * Minimum value allowed for the input
   */
  min: _react.PropTypes.number,
  /**
   * Name of the input attribute
   */
  name: _react.PropTypes.string,
  /**
   * Function to trigger when user selects one of the radio buttons
   */
  onChange: _react.PropTypes.func,
  /**
   * Whether the input is required
   */
  required: _react.PropTypes.bool,
  /**
   * Incremental steps for increasing and descreasing the value
   */
  step: _react.PropTypes.number,
  /**
   * The value of the input element
   */
  value: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number])
};

var defaultProps = {
  attrs: {},
  defaultValue: undefined,
  error: null,
  help: null,
  isInline: false,
  label: null,
  max: null,
  min: null,
  name: null,
  onChange: function onChange() {},
  required: true,
  step: null,
  value: undefined
};

var NumberField = function NumberField(_ref) {
  var attrs = _ref.attrs,
      error = _ref.error,
      help = _ref.help,
      isInline = _ref.isInline,
      label = _ref.label,
      max = _ref.max,
      min = _ref.min,
      name = _ref.name,
      onChange = _ref.onChange,
      required = _ref.required,
      step = _ref.step,
      value = _ref.value,
      defaultValue = _ref.defaultValue,
      customProps = _objectWithoutProperties(_ref, ['attrs', 'error', 'help', 'isInline', 'label', 'max', 'min', 'name', 'onChange', 'required', 'step', 'value', 'defaultValue']);

  return _react2.default.createElement(
    _Field2.default,
    _extends({
      error: error,
      label: label,
      help: help,
      isInline: isInline
    }, customProps),
    _react2.default.createElement(_Input2.default, _extends({
      type: 'number',
      step: step,
      max: max,
      min: min,
      required: required,
      name: name,
      value: value,
      defaultValue: defaultValue,
      onChange: onChange
    }, attrs))
  );
};

NumberField.propTypes = propTypes;
NumberField.defaultProps = defaultProps;

exports.default = NumberField;