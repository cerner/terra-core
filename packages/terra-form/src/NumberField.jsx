import React from 'react';
import PropTypes from 'prop-types';

import Field from './Field';
import Input from './Input';

const propTypes = {
  /**
   * The defaultValue of the input element. Use this to create an uncontrolled input
   */
  defaultValue: PropTypes.number,
  /**
   * Error message for when the input is invalid
   */
  error: PropTypes.node,
  /**
   *  Help element to display with the input
   */
  help: PropTypes.node,
  /**
   * Additional attributes to attach to the number input
   */
  inputAttrs: PropTypes.object,
  /**
   * Whether the field is inline
   */
  isInline: PropTypes.bool,
  /**
   * Label of the input
   */
  label: PropTypes.node,
  /**
   * Attributes to attach to the label
   */
  labelAttrs: PropTypes.object,
  /**
   * Maximum value allowed for the input
   */
  max: PropTypes.number,
  /**
   * Minimum value allowed for the input
   */
  min: PropTypes.number,
  /**
   * Name of the input attribute
   */
  name: PropTypes.string,
  /**
   * Function to trigger when user changes the input value. Provide a function to create a controlled input.
   */
  onChange: PropTypes.func,
  /**
   * Whether the input is required
   */
  required: PropTypes.bool,
  /**
   * Incremental steps for increasing and descreasing the value
   */
  step: PropTypes.number,
  /**
   * The defaultValue of the input element. Use this to create an uncontrolled input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

const defaultProps = {
  defaultValue: undefined,
  error: null,
  help: null,
  inputAttrs: {},
  isInline: false,
  label: null,
  labelAttrs: {},
  max: null,
  min: null,
  name: null,
  onChange: undefined,
  required: false,
  step: null,
  value: undefined,
};

const NumberField = (
  {
    error,
    help,
    inputAttrs,
    isInline,
    label,
    labelAttrs,
    max,
    min,
    name,
    onChange,
    required,
    step,
    value,
    defaultValue,
    ...customProps
  },
) => (
  <Field
    error={error}
    label={label}
    labelAttrs={labelAttrs}
    help={help}
    isInline={isInline}
    required={required}
    {...customProps}
  >
    <Input
      type="number"
      step={step}
      max={max}
      min={min}
      required={required}
      name={name}
      value={value}
      defaultValue={defaultValue}
      onChange={onChange}
      {...inputAttrs}
    />
  </Field>
);

NumberField.propTypes = propTypes;
NumberField.defaultProps = defaultProps;

export default NumberField;
