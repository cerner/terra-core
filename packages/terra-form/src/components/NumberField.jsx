// eslint-disable react/jsx-boolean-value

import React, { PropTypes } from 'react';

import Field from './Field';
import Input from './Input';

const propTypes = {
  /**
   * Additional Attributes to attach to the number input
   */
  attrs: PropTypes.object,
  /**
   * The defaultValue of the input element
   */
  defaultValue: PropTypes.number,
  /**
   * Error message for when the input is invalid
   */
  error: PropTypes.node,
  /**
   *  Help element to display with the input
   */
  help: PropTypes.node,
  /**
   * Whether the field is inline
   */
  isInline: PropTypes.bool,
  /**
   * Label of the input
   */
  label: PropTypes.node,
  /**
   * Maximum value allowed for the input
   */
  max: PropTypes.number,
  /**
   * Minimum value allowed for the input
   */
  min: PropTypes.number,
  /**
   * Name of the input attribute
   */
  name: PropTypes.string,
  /**
   * Function to trigger when user selects one of the radio buttons
   */
  onChange: PropTypes.func,
  /**
   * Whether the input is required
   */
  required: PropTypes.bool,
  /**
   * Incremental steps for increasing and descreasing the value
   */
  step: PropTypes.number,
  /**
   * The value of the input element
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

const defaultProps = {
  attrs: {},
  defaultValue: undefined,
  error: null,
  help: null,
  isInline: false,
  label: null,
  max: null,
  min: null,
  name: null,
  onChange: () => {},
  required: true,
  step: null,
  value: undefined,
};

const NumberField = (
  {
    attrs,
    error,
    help,
    isInline,
    label,
    max,
    min,
    name,
    onChange,
    required,
    step,
    value,
    defaultValue,
    ...customProps
  },
) => (
  <Field
    error={error}
    label={label}
    help={help}
    isInline={isInline}
    {...customProps}
  >
    <Input
      type="number"
      step={step}
      max={max}
      min={min}
      required={required}
      name={name}
      value={value}
      defaultValue={defaultValue}
      onChange={onChange}
      {...attrs}
    />
  </Field>
);


NumberField.propTypes = propTypes;
NumberField.defaultProps = defaultProps;

export default NumberField;
