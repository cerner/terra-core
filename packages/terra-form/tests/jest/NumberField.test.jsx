import React from 'react';
import NumberField from '../../src/NumberField';

it('should render a default component', () => {
  const input = <NumberField />;
  const wrapper = shallow(input);
  expect(wrapper).toMatchSnapshot();
});

it('should render a NumberField when all the possible props are passed into it', () => {
  const input = (
    <NumberField
      label="Sales Tax Rate"
      labelAttrs={{ className: 'healtheintent-legend' }}
      name="sales_tax_rate"
      error="This field is required"
      help="Your county's office may have this information"
      max={1}
      min={0}
      step={0.1}
      inputAttrs={{ className: 'cernerConsumer-application' }}
      inputId="tax-rate"
      isInline
      required
    />);

  const wrapper = shallow(input);
  expect(wrapper).toMatchSnapshot();
});

it('should render as uncontrolled when just a default value is passed into the NumberField', () => {
  const input = <NumberField defaultValue={2} />;
  const wrapper = shallow(input);
  expect(wrapper).toMatchSnapshot();
});

it('should render as controlled when just a default value is passed into the NumberField', () => {
  const input = <NumberField value={2} onChange={() => {}} />;
  const wrapper = mount(input);
  expect(wrapper).toMatchSnapshot();
});
