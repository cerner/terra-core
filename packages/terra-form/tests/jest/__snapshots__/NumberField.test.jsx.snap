exports[`test should render a NumberField when all the possible props are passed into it 1`] = `
<Field
  error="This field is required"
  help="Your county\'s office may have this information"
  htmlFor="tax-rate"
  isInline={true}
  label="Sales Tax Rate"
  labelAttrs={
    Object {
      "className": "healtheintent-legend",
    }
  }
  required={true}>
  <Input
    className="cernerConsumer-application"
    id="tax-rate"
    max={1}
    min={0}
    name="sales_tax_rate"
    required={true}
    step={0.1}
    type="number" />
</Field>
`;

exports[`test should render a default component 1`] = `
<Field
  error={null}
  help={null}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  required={false}>
  <Input
    max={null}
    min={null}
    name={null}
    required={false}
    step={null}
    type="number" />
</Field>
`;

exports[`test should render as controlled when just a default value is passed into the NumberField 1`] = `
<NumberField
  error={null}
  help={null}
  inputAttrs={Object {}}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  max={null}
  min={null}
  name={null}
  onChange={[Function]}
  required={false}
  step={null}
  value={2}>
  <Field
    error={null}
    help={null}
    isInline={false}
    label={null}
    labelAttrs={Object {}}
    required={false}>
    <div
      className="terra-Form-field">
      <Input
        max={null}
        min={null}
        name={null}
        onChange={[Function]}
        required={false}
        step={null}
        type="number"
        value={2}>
        <input
          className="terra-Form-input"
          max={null}
          min={null}
          name={null}
          onChange={[Function]}
          required={false}
          step={null}
          type="number"
          value={2} />
      </Input>
    </div>
  </Field>
</NumberField>
`;

exports[`test should render as uncontrolled when just a default value is passed into the NumberField 1`] = `
<Field
  error={null}
  help={null}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  required={false}>
  <Input
    defaultValue={2}
    max={null}
    min={null}
    name={null}
    required={false}
    step={null}
    type="number" />
</Field>
`;
